servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/versionstore/ObservingVersionStore.java

```diff
@@ -0,0 +1,285 @@
+import org.projectnessie.versioned.RefLogDetails;
+import org.projectnessie.versioned.RefLogNotFoundException;
+import org.projectnessie.versioned.ReferenceAlreadyExistsException;
+import org.projectnessie.versioned.ReferenceAssignedResult;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.ReferenceCreatedResult;
+import org.projectnessie.versioned.ReferenceDeletedResult;
+import org.projectnessie.versioned.ReferenceInfo;
+import org.projectnessie.versioned.ReferenceNotFoundException;
+import org.projectnessie.versioned.RelativeCommitSpec;
+import org.projectnessie.versioned.RepositoryInformation;
+import org.projectnessie.versioned.VersionStore;
+import org.projectnessie.versioned.paging.PaginationIterator;
+
+@Singleton
+public class ObservingVersionStore implements VersionStore {
+  private final VersionStore delegate;
+
+  private static final String PREFIX = "nessie.versionstore.request";
+
+  private static final String TAG_REF = "ref";
+  private static final String TAG_BRANCH = "branch";
+  private static final String TAG_HASH = "hash";
+  private static final String TAG_EXPECTED_HASH = "expected-hash";
+  private static final String TAG_TARGET_HASH = "target-hash";
+  private static final String TAG_FROM = "from";
+  private static final String TAG_TO = "to";
+
+  public ObservingVersionStore(@NotObserved VersionStore delegate) {
+    this.delegate = delegate;
+  }
+
+  @Nonnull
+  @WithSpan
+  @Override
+  @Counted(PREFIX)
+  @Timed(value = PREFIX, histogram = true)
+  public RepositoryInformation getRepositoryInformation() {
+    return delegate.getRepositoryInformation();
+  }
+
+  @WithSpan
+  @Override
+  @Counted(PREFIX)
+  @Timed(value = PREFIX, histogram = true)
+  public Hash hashOnReference(
+      @SpanAttribute(TAG_REF) NamedRef namedReference,
+      @SpanAttribute(TAG_HASH) Optional<Hash> hashOnReference,
+      List<RelativeCommitSpec> relativeLookups)
+      throws ReferenceNotFoundException {
+    return delegate.hashOnReference(namedReference, hashOnReference, relativeLookups);
+  }
+
+  @Nonnull
+  @Override
+  public Hash noAncestorHash() {
+    return delegate.noAncestorHash();
+  }
+
+  @WithSpan
+  @Override
+  @Counted(PREFIX)
+  @Timed(value = PREFIX, histogram = true)
+  public CommitResult<Commit> commit(
+      @SpanAttribute(TAG_BRANCH) @Nonnull BranchName branch,
+      @SpanAttribute(TAG_HASH) @Nonnull Optional<Hash> referenceHash,
+      @Nonnull CommitMeta metadata,
+      @Nonnull List<Operation> operations,
+      @Nonnull VersionStore.CommitValidator validator,
+      @Nonnull BiConsumer<ContentKey, String> addedContents)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    return delegate.commit(branch, referenceHash, metadata, operations, validator, addedContents);
+  }
+
+  @WithSpan
+  @Override
+  @Counted(PREFIX)
+  @Timed(value = PREFIX, histogram = true)
+  public MergeResult<Commit> transplant(TransplantOp transplantOp)
```

Comment:
Here, and in a few other places, we are losing the granularity that we used to have for span attributes. The old code had this:

```java
 b.setAttribute(TAG_TARGET_BRANCH, safeRefName(transplantOp.toBranch()))
                        .setAttribute(TAG_HASH, safeToString(transplantOp.expectedHash()))
                        .setAttribute(
                            TAG_TRANSPLANTS, safeSize(transplantOp.sequenceToTransplant())),
```

Are we OK with this? Alternatively, you could:

1. create the span manually (but I suppose you don't want to, because if the tracing extension is not present at runtime, this would result in a linkage error – annotations are much more flexible);
2. call a delegate method with parameters for all the fields in `TransplantOp` that we want span attributes for.

Wdyt?

Reply:
Yea - the drawback is that we lose a few attributes. However, I think we didn't really use those attributes.

Reply:
Sure, we can add delegate methods - if there's demand for that.